### ISSUE TITLE
Collection scanning extremely slow and unreliable for external hard drive

### ISSUE THREAD

**Issue by eignatenkov** [strawberrymusicplayer/strawberry#867]
I have one music folder (160GB) on my laptop hard drive and another one on the external USB hard drive (400GB). The first part is being scanned pretty quickly, in a couple of seconds, the second one takes more than an hour to be updated. To compare, Foobar scans both folders in several seconds. And then, sometimes, when I add new folders, they don't appear in the collection, even after I do "Update changed collection folders". And, obviously, rescanning the whole library completely is not a very good option, since it takes more than an hour.

I tried to remove the strawberry files from .local/share/strawberry, remove all folders from the collection and then add them again, it did not help, still slow.

Ubuntu 21.10, Strawberry 1.0

**Comment by eignatenkov** [comment]
USB drive only: 400GB, 35K files in the folder, 28K mp3s, 1700 flacs. Filesystem vfat, disk SP PHD U3.
Removed the folder from foobar library, turned the disc on and off, added folder back to foobar's library. Took 10 minutes. Strawberry -- 2 hours, order of magnitude slower.

**Comment by jonaski** [comment]
Do you have the fingerprinting feature in the collection settings on or off?
If it's off, then 2 hours is not normal, I have 24K songs and I think it usually takes 15 minutes or so without disk cache and only a few seconds with disk cache. With fingerprinting it will be much slower.
Foobar is only available on Windows so I assume it's not the same computer you use strawberry on as foobar?
But it's the same disk?
What CPU does the computer have?
Strawberry uses idle I/O and thread priority for the thread running the file scan, but on a modern machine it should normally not be a problem as long as the CPU isn't very busy, but I definitely see that it is very slow if I compile on the machine while Strawberry is scanning the collection.

**Comment by eignatenkov** [comment]
Fingerprinting is off.
Foobar is available in snap, it's on the same computer.
CPU is 8 cores x 1.8GHz. I don't see it being extremely busy when strawberry is scanning the collection. Looks more like it's slow with the reading from the hard drive.

**Comment by eignatenkov** [comment]
What's more, when I disconnect the hard drive (for instance when I go home from work) and then don't plug it back in fast enough, Strawberry seems to realize that one of the music folders is gone and updates the library (as in, hides the music that is stored on the external drive). And then when I finally (and I'm talking about several hours, not days, it just happened today again) connect it again, it seems that it doesn't remember a thing about it, starts scanning it right from the start, so it's another two hours until it's done.

Liked everything about the player but when it's about the music library and my external drive it's like entering the world of constant pain.

**Comment by jonaski** [comment]
I added advanced section to the collection settings where you can configure I/O priority, tread priority and workers (https://github.com/strawberrymusicplayer/strawberry/commit/e31c9d74fa8174c6ad1913e10027cee891f9475f).
Try setting I/O priority to Auto, thread priority to Low and tagreader workers to 1.

**Comment by eignatenkov** [comment]
I'm sorry, I've been away for a while. What's the right way to test the change? Should I build from the current master state manually?

**Comment by jonaski** [comment]
You can use the "unstable" PPA: https://launchpad.net/~jonaski/+archive/ubuntu/strawberry-unstable
or manually install the latest from https://builds.strawberrymusicplayer.org/
I planned to make a new release this weekend but I found some bugs I wanted to fix first so it's a bit delayed.

**Comment by eignatenkov** [comment]
Okay, I tried the new settings, they don't seem to improve the scanning speed. I also switched to a new laptop, no improvement, scanning from the internal drive is blazing fast, but with the external one it is as slow as it was.

**Comment by jonaski** [comment]
I'm not familiar with how foobar works but if Foobar scans the folder in a few seconds it can't possible read data from all the files,
but only walk through the directories and index the files. If it shows metadata then I assume it has to use some sort of cache from previous scans that it uses.
The problem in Strawberry is that the watcher doesn't have any relation to whether the files have been marked as unavailable and are still in the database, so it simply scans the files that are not in the database, or marked as unavailable in the database.
I guess this could be improved so that when songs are marked unavailable they aren't re-scanned when they re-appear unless the mtime changes. But of course this does not solve the initial scan time when a new folder is added.

**Comment by jonaski** [comment]
Closing this as there doesn't seem to be an fix for this, other than the hdd being slow.

**Comment by Jellby** [comment]
I'm seeing something similar. While it's true that the external drive may be slow, I believe the scanning could be made faster. For instance, an update with rsync is much faster. I wonder if the database could store file timestamps and/or checksums to avoid unnecessary updates.

**Comment by jonaski** [comment]
Of course we have timestamps in the database. We are talking about initial scan here. And like I already mentioned, songs that re-appear are re-read independent of timestamp, so that could possibly be improved, but it doesn't fix the underlying speed of the reading the files.
If you think it could be improved, a fix is always welcome.

It might be worth re-testing with TagLib 1.13 which includes this fix and might actually improve the speed: https://github.com/taglib/taglib/commit/2e90ec0ef06d45c18bbe9725e766f1d968f7441c

**Comment by 4oo4** [comment]
Not sure if it's the same issue but I'm also having performance issues reading files from an NFS share. 

Client and server are both Debian Bullseye, I'm using the settings mentioned here, but I can't seem to get anything better than 5 seconds per song.

https://github.com/strawberrymusicplayer/strawberry/issues/867#issuecomment-1025060107

I haven't encountered this with other applications, client and server are on the same switch and usually I can get about 120-150 MBps of throughput for nfs. I'm currently using 1.0.9 from the bullseye-backports package, but I'll try 1.0.10 next, and then also possibly try building with the updated taglib.

What's really interesting is that it takes exactly 5 second per song, so it could also be something weird with my setup like nfs file locking coming into play.

```
Nov 16 18:24:27 localhost org.strawberrymusicplayer.strawberry.desktop[31913]: 18:24:27.653 DEBUG MessageReply<MessageType>:90     Releasing ID 121 (finished)
Nov 16 18:24:27 localhost org.strawberrymusicplayer.strawberry.desktop[31913]: 18:24:27.653 DEBUG _MessageReplyBase:33             Acquired ID 121
Nov 16 18:24:27 localhost org.strawberrymusicplayer.strawberry.desktop[31913]: 18:24:27.653 DEBUG _MessageReplyBase:31             Waiting on ID 122
Nov 16 18:24:27 localhost org.strawberrymusicplayer.strawberry.desktop[31928]: 18:24:27.654 DEBUG TagReaderTagLib:151              Checking for valid file "/path/to/file.flac"
Nov 16 18:24:32 localhost org.strawberrymusicplayer.strawberry.desktop[31913]: 18:24:32.773 DEBUG MessageReply<MessageType>:90     Releasing ID 122 (finished)
Nov 16 18:24:32 localhost org.strawberrymusicplayer.strawberry.desktop[31913]: 18:24:32.773 DEBUG _MessageReplyBase:33             Acquired ID 122
Nov 16 18:24:32 localhost org.strawberrymusicplayer.strawberry.desktop[31913]: 18:24:32.774 DEBUG _MessageReplyBase:31             Waiting on ID 123
Nov 16 18:24:32 localhost org.strawberrymusicplayer.strawberry.desktop[31928]: 18:24:32.774 DEBUG TagReaderTagLib:151              Checking for valid file "/path/to/file.flac"
Nov 16 18:24:37 localhost org.strawberrymusicplayer.strawberry.desktop[31913]: 18:24:37.897 DEBUG MessageReply<MessageType>:90     Releasing ID 123 (finished)
Nov 16 18:24:37 localhost org.strawberrymusicplayer.strawberry.desktop[31913]: 18:24:37.897 DEBUG _MessageReplyBase:33             Acquired ID 123
Nov 16 18:24:37 localhost org.strawberrymusicplayer.strawberry.desktop[31913]: 18:24:37.898 DEBUG _MessageReplyBase:31             Waiting on ID 124
Nov 16 18:24:37 localhost org.strawberrymusicplayer.strawberry.desktop[31928]: 18:24:37.898 DEBUG TagReaderTagLib:151              Checking for valid file "/path/to/file.flac"
Nov 16 18:24:43 localhost org.strawberrymusicplayer.strawberry.desktop[31913]: 18:24:43.013 DEBUG MessageReply<MessageType>:90     Releasing ID 124 (finished)
Nov 16 18:24:43 localhost org.strawberrymusicplayer.strawberry.desktop[31913]: 18:24:43.013 DEBUG _MessageReplyBase:33             Acquired ID 124
Nov 16 18:24:43 localhost org.strawberrymusicplayer.strawberry.desktop[31913]: 18:24:43.013 DEBUG _MessageReplyBase:31             Waiting on ID 125
Nov 16 18:24:43 localhost org.strawberrymusicplayer.strawberry.desktop[31928]: 18:24:43.013 DEBUG TagReaderTagLib:194              Reading tags from "/path/to/file.flac"
Nov 16 18:24:48 localhost org.strawberrymusicplayer.strawberry.desktop[31913]: 18:24:48.133 DEBUG MessageReply<MessageType>:90     Releasing ID 125 (finished)
Nov 16 18:24:48 localhost org.strawberrymusicplayer.strawberry.desktop[31913]: 18:24:48.133 DEBUG _MessageReplyBase:33             Acquired ID 125
Nov 16 18:24:48 localhost org.strawberrymusicplayer.strawberry.desktop[31913]: 18:24:48.133 DEBUG CollectionWatcher:654            "/path/to/file.flac" is new.
Nov 16 18:24:48 localhost org.strawberrymusicplayer.strawberry.desktop[31913]: 18:24:48.133 DEBUG _MessageReplyBase:31             Waiting on ID 126
Nov 16 18:24:48 localhost org.strawberrymusicplayer.strawberry.desktop[31928]: 18:24:48.133 DEBUG TagReaderTagLib:194              Reading tags from "/path/to/file.flac"
Nov 16 18:24:53 localhost org.strawberrymusicplayer.strawberry.desktop[31913]: 18:24:53.253 DEBUG MessageReply<MessageType>:90     Releasing ID 126 (finished)
Nov 16 18:24:53 localhost org.strawberrymusicplayer.strawberry.desktop[31913]: 18:24:53.253 DEBUG _MessageReplyBase:33             Acquired ID 126
Nov 16 18:24:53 localhost org.strawberrymusicplayer.strawberry.desktop[31913]: 18:24:53.253 DEBUG CollectionWatcher:654            "/path/to/file.flac" is new.
Nov 16 18:24:53 localhost org.strawberrymusicplayer.strawberry.desktop[31913]: 18:24:53.253 DEBUG _MessageReplyBase:31             Waiting on ID 127
Nov 16 18:24:53 localhost org.strawberrymusicplayer.strawberry.desktop[31928]: 18:24:53.253 DEBUG TagReaderTagLib:194              Reading tags from "/path/to/file.flac"
Nov 16 18:24:58 localhost org.strawberrymusicplayer.strawberry.desktop[31913]: 18:24:58.373 DEBUG MessageReply<MessageType>:90     Releasing ID 127 (finished)
Nov 16 18:24:58 localhost org.strawberrymusicplayer.strawberry.desktop[31913]: 18:24:58.373 DEBUG _MessageReplyBase:33             Acquired ID 127
Nov 16 18:24:58 localhost org.strawberrymusicplayer.strawberry.desktop[31913]: 18:24:58.373 DEBUG CollectionWatcher:654            "/path/to/file.flac" is new.
Nov 16 18:24:58 localhost org.strawberrymusicplayer.strawberry.desktop[31913]: 18:24:58.373 DEBUG _MessageReplyBase:31             Waiting on ID 128
Nov 16 18:24:58 localhost org.strawberrymusicplayer.strawberry.desktop[31928]: 18:24:58.373 DEBUG TagReaderTagLib:194              Reading tags from "/path/to/file.flac"
```

**Comment by jonaski** [comment]
> Not sure if it's the same issue but I'm also having performance issues reading files from an NFS share.
> 
> Client and server are both Debian Bullseye, I'm using the settings mentioned here, but I can't seem to get anything better than 5 seconds per song.
> 
5 seconds per file/song is crazy. Sounds to me like either your HDD or network is slow, if you get the same result with local files, it could be that Debian has a buggy version of TagLib. Make sure you use at least TagLib 1.12, preferable 1.13. See if they have a test utility for TagLib, that helps nail down to where the issue is.

I tested adding a NFS directory here just now with circa 25.000 FLAC songs, it finished scanning in approximately 30 minutes. It scans each file in a few milliseconds. If it's the same drive but local, it's scanned in half the time, circa 10-15 minutes. This is with TagLib 1.12 on openSUSE.

**Comment by jonaski** [comment]
TagLib has a test utility called `tagreader`, I assume debian must have package for it.
You can use: `time tagreader` <filename>

**Comment by 4oo4** [comment]
Thanks, I'll experiment with `taglib` and see if I can narrow it down, and open a new issue if I find something. When you tested with NFS, was it also version 4?

I'm doubtful it's network or HDD, since usually my nfs speeds are 100+ MBps and client and server are on the same switch, and files are being served from a zfs pool with 7200 RPM HDDs and an M.2 SSD cache drive. This is the first application that I've seen these types of performance issues with.

Thanks
